src/storage/store/engine.rs

```diff
@@ -1243,31 +1318,62 @@ impl ShardEngine {
             version,
         )?;
 
-        // Check that the user has a custody address
-        self.stores
-            .onchain_event_store
-            .get_id_register_event_by_fid(message_data.fid)
-            .map_err(|_| MessageValidationError::MissingFid)?
-            .ok_or(MessageValidationError::MissingFid)?;
+        // 1. Check that the user has a custody address
+        if !txn_onchain_state
+            .id_registers
+            .contains_key(&message_data.fid)
+        {
+            // If not in the temp cache, fall back to the DB
+            self.stores
+                .onchain_event_store
+                .get_id_register_event_by_fid(message_data.fid)
+                .map_err(|_| MessageValidationError::MissingFid)?
+                .ok_or(MessageValidationError::MissingFid)?;
+        }
 
-        // Check that signer is valid
-        self.stores
-            .onchain_event_store
-            .get_active_signer(message_data.fid, message.signer.clone())
-            .map_err(|_| MessageValidationError::MissingSigner)?
-            .ok_or(MessageValidationError::MissingSigner)?;
+        // 2. Check that the user has a valid signer
+        match txn_onchain_state.get_effective_signer_state(message_data.fid, &message.signer) {
```

Comment:
Why do this vs passing the txn batch to `get_active_signer`? I think this should be the same, but I'm not sure if there are some edges around event ordering where this map may not match what's actually in the db.

Reply:
So i put it here to support this (admittedly contrived) scenario - If there is a SignerRemove message, and then a CastAdd that uses the signer in the same batch, then the CastAdd should fail. The `get_effective_signer_state` handles the signer removes, but it would be weird to have that be in get_signer_add. 

Don't have a strong opinion here, so if you think this is too contrived, I can switch it into get_signer_add

Reply:
I have the same question as Sanjay here. I thought this would be implemented by threading the currently active txn through to more stores. 

I'm not sure I understand what the issue is with this approach even in the scenario you described. `get_active_signer` should account for signer removes too.